plat/common/lcpu.c

```diff
@@ -48,3 +182,336 @@ void ukplat_lcpu_halt_to(__nsec until)
+		return rc;
+	}
+
+	return 0;
+}
+
+__lcpuidx lcpu_idx(struct lcpu *lcpu)
+{
+	UK_ASSERT(lcpu_count > 0);
+	UK_ASSERT(lcpu >= lcpus); /* Defined for elements of same array */
+
+	return (__lcpuidx) (lcpu - lcpus);
+}
+
+#ifdef CONFIG_UKPLAT_LCPU_IDISIDX
+/* For many VMMs the CPU ID is simply a sequentially increasing number with
+ * the BSP always at index 0. This is essentially the definition of __lcpuidx
+ * so in this case, we can simply use the ID as IDX.
+ */
+__lcpuidx ukplat_lcpu_idx(void)
+{
+	__lcpuidx this_cpu_idx = (__lcpuidx) ukplat_lcpu_id();
+
+	UK_ASSERT(this_cpu_idx < lcpu_count);
+	UK_ASSERT(lcpus[this_cpu_idx].idx == this_cpu_idx);
+	return this_cpu_idx;
+}
+#else /* CONFIG_UKPLAT_LCPU_IDISIDX */
+__lcpuidx ukplat_lcpu_idx(void)
+{
+	__lcpuid this_cpu_id  = ukplat_lcpu_id();
+	__u32 i, this_cpu_idx = (__u32) (-1);
+
+	/** TODO: Until we have a better way just do a linear search */
+	for (i = 0; i < lcpu_count; i++)
+		if (lcpus[i].id == this_cpu_id) {
+			UK_ASSERT(lcpus[i].idx == (__lcpuidx) i);
+			this_cpu_idx = i;
+			break;
+		}
+
+	UK_ASSERT(this_cpu_idx != (__u32) (-1));
+
+	return (__lcpuidx) this_cpu_idx;
+}
+#endif /* CONFIG_UKPLAT_LCPU_IDISIDX */
+
+__u32 ukplat_lcpu_count(void)
+{
+	return lcpu_count;
+}
+
+int ukplat_lcpu_start(__lcpuidx lcpuidx[], unsigned int num, void *sp[],
+		      ukplat_lcpu_entry_t entry[], unsigned long flags)
+{
+	__lcpuid this_cpu_id = ukplat_lcpu_id();
+	struct lcpu *lcpu;
+	unsigned int i, n = 0;
+	const int new = LCPU_STATE_INIT;
+	int old;
+	int rc, err = 0;
+
+	UK_ASSERT(sp);
+
+	lcpu_lcpuidx_list_foreach(lcpuidx, num, i, lcpu) {
+		if (lcpu->id == this_cpu_id) {
+			/* If the caller did not supply an index array, we
+			 * assume that we do not have parameters and a stack
+			 * for the executing CPU. Otherwise, i.e., the caller
+			 * explicitly put the executing CPU in. We still
+			 * ignore it but need to skip the parameters.
+			 */
+			if (lcpuidx)
+				n++;
+
+			continue;
+		}
+
+retry:
+		old = UK_READ_ONCE(lcpu->state);
+
+		/* We ignore CPUs that are already started */
+		if (unlikely(old != LCPU_STATE_OFFLINE)) {
+			uk_pr_warn("Failed to start CPU 0x%lx: not offline\n",
+				   lcpu->id);
+
+			/* Skip CPU and its arguments*/
+			n++;
+			continue;
+		}
+
+		/* Try to acquire the CPU for initialization. If another thread
+		 * was faster, we will return to the state comparison and
+		 * report that the CPU is not offline.
+		 */
+		if (ukarch_compare_exchange_sync((int *)&lcpu->state, old,
+						 new) != new)
+			goto retry;
+
+		UK_ASSERT(lcpu->state == LCPU_STATE_INIT);
+
+		/* Setup startup arguments.
+		 * Since we are ignoring the executing CPU, we must keep a
+		 * separate counter to index the arguments.
+		 */
+		lcpu->s_args.entry = (entry && entry[n]) ?
+			entry[n] : lcpu_entry_default;
+		lcpu->s_args.stackp = sp[n];
+
+		n++;
+
+		/* Ensure that the startup arguments have been written back
+		 * before issuing the startup IPI
```

Comment:
Not necessarily IPI 🙂 PSCI is a call to the firmware.

Reply:
👍🏻